src/ordered.jl

```diff
@@ -10,30 +10,32 @@ Optionally, this final threshold map can be inverted by selecting `invert_map=tr
 function ordered_dithering(
     img::AbstractMatrix{<:Gray}, mat::AbstractMatrix; to_linear=false, invert_map=false
 )::BitMatrix
-    # Create full threshold map by repeating threshold matrix `mat` in x and y directions
-    h, w = size(img)
-    threshold = tile_matrix(mat, h, w)
-
-    # Optionally cast to linear colorspace
-    _img = copy(img)
-    to_linear && srgb2linear!(_img)
-
-    # Optionally invert map ∈ [0, 1]
-    invert_map && (threshold .= 1 .- threshold)
-
-    return _img .> threshold
-end
-
-"""
-    tile_matrix(mat, h, w)
-
-Repeatedly tile a smaller matrix `mat` to fill out an image of height `h` and width `w`.
-"""
-function tile_matrix(mat, h, w)
-    h_mat, w_mat = size(mat)
-    repeat_rows = ceil(Int, h / h_mat)
-    repeat_cols = ceil(Int, w / w_mat)
-    return view(repeat(mat, repeat_rows, repeat_cols), 1:h, 1:w) # view matching image dims
+    # eagerly promote to the same eltype to make for-loop faster
+    FT = floattype(eltype(img))
+    if invert_map
+        mat = @. FT(1 - mat)
+    else
+        mat = FT.(mat)
+    end
+    linear_fun = to_linear ? srgb2linear : identity
+    img = @. FT.(linear_fun.(img))
+
+    out = BitMatrix(undef, size(img)...)
+    # TODO: add Threads.@threads to this for loop further improves the performances
+    #       but it has unidentified memory allocations
+    @inbounds for R in TileIterator(axes(img), size(mat))
+        mat_size = map(length, R)
+        if mat_size == size(mat)
+            # `mappedarray` creates a readonly wrapper with lazy evaluation with `srgb2linear`
+            # so that original `img` data is not modified.
```

Comment:
lol, this is an outdated comment; it was `img = mappedarray(x->FT(linear_fun(x)), img)` when I wrote this, but then I realize that it has some type inference issues in the new MappedArrays version.